### ISSUE TITLE
Setting :application to a Proc causes cap to crash

### ISSUE THREAD

**Issue by mattbrictson** [capistrano/capistrano#1681]
The logic used to validate `:application` expects a string value and will cause an exception to be raised otherwise. This prevents us from setting `:application` using a Proc, as originally reported here: https://github.com/capistrano/documentation/pull/186

#### Steps to reproduce

1. In `deploy.rb`, configure `:application` with a lambda, like this: `set :application, -> { "myapp" }`
2. Run `cap production deploy`

#### Expected behaviour

The deploy should succeed with `myapp` used as the application name.

#### Actual behaviour

```
cap aborted!
NoMethodError: undefined method `gsub' for #<Proc:0x007f9aa311dd80@config/deploy.rb:1 (lambda)>
capistrano-3.5.0/lib/capistrano/defaults.rb:2:in `block in <top (required)>'
capistrano-3.5.0/lib/capistrano/configuration/variables.rb:126:in `block in invoke_validations'
capistrano-3.5.0/lib/capistrano/configuration/variables.rb:125:in `each'
capistrano-3.5.0/lib/capistrano/configuration/variables.rb:125:in `invoke_validations'
capistrano-3.5.0/lib/capistrano/configuration/variables.rb:34:in `set'
config/deploy.rb:1:in `<top (required)>'
```

**Comment by JohnSmall** [comment]
How come this is closed? It's definitely broken in version 3.5. I just upgraded from 3.4 to 3.5 and I get this error, even when I do cap -T, not even deploying.

Do we have to wait till the next bug fix release?

**Comment by mattbrictson** [comment]
This issue was closed by PR #1711, which was merged on June 21. That means it is fixed in master and will be included in the next release. This is typical of most open source projects on GitHub, where issues are closed once they are fixed in the main development branch.

We don't have a fixed release schedule for Capistrano, so I can't promise when the next version will be published, but in the meantime you can point your Gemfile at the master branch:

```ruby
gem "capistrano", :github => "capistrano/capistrano"
```

**Comment by mattbrictson** [comment]
The fix in #1711, as reported by @deekthesqueak, is incomplete.

Specifically:

```ruby
# Using a block works
set(:application) { ... }

# But using a proc or lambda still causes a crash
set :application, -> { ... }
# => NoMethodError: undefined method `gsub' for #<Proc>
```

**Comment by mattbrictson** [comment]
Anyone affected by this bug please take a look at PR #1739 and let me know if it is a good fix for you. Thanks!
